src/st-trace/trace.c

```diff
@@ -0,0 +1,610 @@
+    puts("  -v, --verbose         Specify a generally verbose logging");
+    puts("  -cXX, --clock=XX      Specify the core frequency in MHz");
+    puts("  -tXX, --trace=XX      Specify the trace frequency in Hz");
+    puts("  -n, --no-reset        Do not reset board on connection");
+    puts("  -sXX, --serial=XX     Use a specific serial number");
+    puts("  -f, --force           Ignore most initialization errors");
+}
+
+bool parse_options(int argc, char** argv, st_settings_t *settings) {
+
+    static struct option long_options[] = {
+        {"help",     no_argument,       NULL, 'h'},
+        {"version",  no_argument,       NULL, 'V'},
+        {"verbose",  optional_argument, NULL, 'v'},
+        {"clock",    required_argument, NULL, 'c'},
+        {"trace",    required_argument, NULL, 't'},
+        {"no-reset", no_argument,       NULL, 'n'},
+        {"serial",   required_argument, NULL, 's'},
+        {"force",    no_argument,       NULL, 'f'},
+        {0, 0, 0, 0},
+    };
+    int option_index = 0;
+    int c;
+    bool error = false;
+
+    settings->show_help = false;
+    settings->show_version = false;
+    settings->logging_level = DEFAULT_LOGGING_LEVEL;
+    settings->core_frequency = 0;
+    settings->trace_frequency = 0;
+    settings->reset_board = true;
+    settings->force = false;
+    settings->serial_number = NULL;
+    ugly_init(settings->logging_level);
+
+    while ((c = getopt_long(argc, argv, "hVv::c:ns:f", long_options, &option_index)) != -1) {
+        switch (c) {
+        case 'h':
+            settings->show_help = true;
+            break;
+        case 'V':
+            settings->show_version = true;
+            break;
+        case 'v':
+            if (optarg) {
+                settings->logging_level = atoi(optarg);
+            } else {
+                settings->logging_level = DEBUG_LOGGING_LEVEL;
+            }
+            ugly_init(settings->logging_level);
+            break;
+        case 'c':
+            settings->core_frequency = atoi(optarg) * 1000000;
+            break;
+        case 't':
+            settings->trace_frequency = atoi(optarg);
+            break;
+        case 'n':
+            settings->reset_board = false;
+            break;
+        case 'f':
+            settings->force = true;
+            break;
+        case 's':
+            settings->serial_number = optarg;
+            break;
+        case '?':
+            error = true;
+            break;
+        default:
+            ELOG("Unknown command line option: '%c' (0x%02x)\n", c, c);
+            error = true;
+            break;
+        }
+    }
+
+    if (optind < argc) {
+        while (optind < argc) {
+            ELOG("Unknown command line argument: '%s'\n", argv[optind++]);
+        }
+        error = true;
+    }
+
+    if (error && !settings->force)
+        return false;
+
+    return true;
+}
+
+static void ConvertSerialNumberTextToBinary(const char* text, char binary_out[STLINK_SERIAL_MAX_SIZE]) {
```

Comment:
all functions have named in _snake_case_, so I think in this file this also should be a rule.